include/wlr/types/wlr_xdg_foreign_v1.h

```diff
@@ -0,0 +1,75 @@
+/*
+ * This an unstable interface of wlroots. No guarantees are made regarding the
+ * future consistency of this API.
+ */
+#ifndef WLR_USE_UNSTABLE
+#error "Add -DWLR_USE_UNSTABLE to enable unstable wlroots features"
+#endif
+
+#ifndef WLR_TYPES_WLR_XDG_FOREIGN_V1_H
+#define WLR_TYPES_WLR_XDG_FOREIGN_V1_H
+
+#include <wayland-server-core.h>
+
+struct wlr_xdg_foreign_v1 {
+	struct {
+		struct wl_global *global;
+		struct wl_list resources;
+		struct wl_list clients;
+	} exporter, importer;
+
+	struct wl_listener display_destroy;
+
+	struct {
+		struct wl_signal destroy;
+	} events;
+
+	void *data;
+};
+
+struct wlr_xdg_exporter_v1 {
+	struct wl_resource *resource;
+
+	struct wl_list link; // wlr_xdg_foreign_v1::exporter::clients
+
+	struct wl_list exports;
+};
+
+struct wlr_xdg_importer_v1 {
+	struct wlr_xdg_foreign_v1 *foreign;
+	struct wl_resource *resource;
+
+	struct wl_list link; // wlr_xdg_foreign_v1::importer::clients
+
+	struct wl_list imports;
+};
+
+struct wlr_xdg_exported_v1 {
+	struct wlr_surface *surface;
+	struct wl_resource *resource;
+
+	struct wl_list link; // wlr_xdg_exporter_v1::exports
+
+	struct wl_listener xdg_surface_unmap;
+
+	struct wl_list imports;
+	char *handle;
+};
+
+struct wlr_xdg_imported_v1 {
+	struct wlr_xdg_exported_v1 *exported;
+	struct wl_resource *resource;
+
+	struct wl_list export_link; // wlr_xdg_exported_v1::imports
+	struct wl_list link; // wlr_xdg_importer_v1::imports
+
+	struct wl_listener child_unmap;
+
+	struct wlr_surface *child;
```

Comment:
To my understanding, the protocol allows multiple `set_parent_of` calls on the same `xdg_imported`, however here you only keep track of one child.

To clarify, I think the following is possible:

1. Create dialogues A and B
2. Import surface C
3. Set C as parent of A
4. Set C as parent of B

And now  both A and B are children of C.

Reply:
This has been replaced with a list of children. It's necessary to reset the parents after the imported resource gets destroyed.